### ISSUE TITLE
Clarify behavior of constructor initializer list augmentations 

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#3237]
The proposal doesn't talk about these directly today, but it hints that they are allowed with this text `The original constructor is a factory constructor and the augmenting constructor has an initializer list.` (which is an error, but that isn't super relevant).

In particular, does an augmenting initializer list _add_ to the existing one, or replace it? If it does _add_ to it, can you access the original initializer list or individual initializer for a specific field? We could for instance say they are "merged" and where there was an original initializer for the same field, the augmenting one can call `augment super` to refer to the original. 

See also https://github.com/dart-lang/language/issues/3230 as the behavior is relevant for that issue.

In general I think I lean towards saying they are only additive, meaning you can't have an initializer for a field that had one originally, and multiple augmentations can't add initializers for the same field. We could always change this to a merging behavior in a non-breaking release in the future, if we had compelling reasons to do so.

**Comment by rrousselGit** [comment]
I would lean toward making initializers additive. And making it an error if an augmented initializer re-initialize an already initialized field.

Then if we later want the "merge" behavior, this wouldn't be breaking.

**Comment by lrhn** [comment]
If going with a "syntactic addition" view, or makes sense to add initializer list entries, parameters (likely optional, can be initializing or super), and fields to initialize.

I'm assuming a library only has to be complete and consistent when all augmentations are applied, so it's fine for an augmentation to add an initializer for a field to the existing constructor, or a field that an existing constructor is already initializing.

So I don't see any problems with adding more initializers to an existing non-redirecting generative constructor.

Or more arguments to the super-constructor call (trailing positional or names only).
Basically anything you could have added to the original, without changing the meaning of what was already written.

**Comment by lrhn** [comment]
I've looked at this again.

Where we augment a method, we really replace it with another method, which is a valid override, and which can choose to call the original instead.

We should do the same for constructors, not allow adding individual parts.

So an augment of a generative constructor should be a full generative constructor, where the `this` (if redirecting) or `super` (if generative) invocation can refer to a pre-augment constructor.

```
augment class C { 
  final int id;
  C(super.x, super.y, this.id) : augment super();
}
```

Then you can augment a generative constructor with any other generative constructor, and a factory constructor with any constructor at all.
You can change which parameters are initializing and which are super, depending on what you need for your constructor.

**Comment by jakemac53** [comment]
Hmm, that is an interesting way to look at the feature, but I don't think it allows for keeping the original initializer list and also adding a new super class + calling a constructor from it? Can you have both an `augment super()` and `super()` call? That doesn't seem feasible to me.

**Comment by lrhn** [comment]
It won't allow that, but should we allow it?
Preserving the initializer list somewhat presumes preserving the parameter list (ok, I can work around that, as long as it has at least the same parameters, then we can probably ignore names of positional parameters), and preserving the parameter list means preserving which parameters are initializing formals and super parameters, and at that point replacing the super constructor gets tricky.
Isn't it better to rewrite everything, if you really need to change what the constructor does? Then either chain to the existing being-augmented constructor, or choose another super-constructor entirely.

**Comment by jakemac53** [comment]
We don't allow messing with parameters to functions - both me and Bob felt it was just too weird/unexpected to do that. It would also be very weird/unexpected to bash entirely over the initializer list - the original one would then have zero value and we would end up with a lint/warning saying to remove it or something.

The model for function augmentations is that only the body is replaced - and in fact the macro APIs only accept a function body for the augmentation, not a parameter list.

I think once you take augmenting parameters out of the equation, the model of merging in the initializers makes a lot more sense. Also, I think it is the most useful model.

**Comment by rrousselGit** [comment]
I still wish there were ways to remove/add parameters and make some optional.
There are many function composition examples needing such a thing. Like currying 

Dart is quite limited in the context of function composition.
I would've hoped macros could solve that.

Otherwise users will start annotating one symbol but have to use a different one instead.
This gets confusing, as the name of the new symbol is often arbitrary. 

Take Riverpod. It's mechanism is in many way similar to:

```ts
// the annotation
function riverpod(callback) {
  const ref = new Ref();
  return (...args) => callback(ref, ...args)
}

// user defined code
@riverpod
function myFunction(ref, {id}): string {
  return `Hello ${id}`;
} 

// used as
const label: string = ref.watch(myFunction({id: 42})) 
```


Being unable to do such composition means that although users defined a `myFunction`, they can't use it. That symbol is useless for them.

In Dart, if they wished to do that, they'd do:

```dart
@riverpod
String $myFunction(Ref ref, {int? id}) {...}
```

Which would generate a `myFunction` without the $.

That feels like the sort of thing augmentation libraries shouldn't need.

**Comment by jakemac53** [comment]
It looks like I just missed this in the spec as well "If the augmenting constructor has any initializers, they are appended to
the original constructor's initializers, but before any original super initializer or original redirecting initializer if there is one.". Closing this.
